### PR TITLE
Prometheus encoding

### PR DESCRIPTION
**Pull request by chris124567**
Continuation of https://github.com/SiaFoundation/renterd/pull/1068

Addressed some of the concerns in the reviews:
- Move `PrometheusMetric` method to original type where possible
- Move `writeResponse` logic to api package
- Improve handling slices of Prometheus marshallable types
- Disable type checking in `jape`

### CODE REVIEW

**Review by ChrisSchinnerl — CHANGES_REQUESTED** [review]

File: api/prometheus.go [hunk 1]
```diff
@@ -0,0 +1,908 @@
...
+		Value: 1,
+	})
+	return
+}
+
+func (host Host) PrometheusMetric() (metrics []prometheus.Metric) {
+	priceTableLabels := map[string]any{
+		"netAddress": host.NetAddress,
+		"uid":        host.PriceTable.UID.String(),
+		"expiry":     host.PriceTable.Expiry.Local().Format("2006-01-02T15:04:05Z07:00"),
```
Comment (peterjan) [inline api/prometheus.go:122]: `nit` use `.Format(time.RFC3339)` so if we ever change time formats a full search also shows this line

File: api/utils.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+import (
+	"fmt"
+	"net/http"
+
+	"go.sia.tech/jape"
+	"go.sia.tech/renterd/internal/prometheus"
+)
+
+func WriteResponse(jc jape.Context, code int, resp any) error {
```
Comment (peterjan) [inline api/utils.go:11]: @ChrisSchinnerl do you think it's worth moving this into the `jape` package?

File: api/utils.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+	var responseFormat string
+	if err := jc.DecodeForm("response", &responseFormat); err != nil {
+		return err
```
Comment (peterjan) [inline api/utils.go:14]: The error returned here is never evaluated, so we probably don't write the proper status code

File: api/utils.go (hunk 2 above)
Comment (peterjan) [inline api/utils.go:11]: The status code is not used

File: api/prometheus.go [hunk 4]
```diff
@@ -0,0 +1,908 @@
...
+}
+
+func (sr BusStateResponse) PrometheusMetric() (metrics []prometheus.Metric) {
+	metrics = append(metrics, prometheus.Metric{
+		Name: "renterd_state",
+		Labels: map[string]any{
+			"network":   sr.Network,
+			"version":   sr.Version,
+			"commit":    sr.Commit,
+			"starttime": sr.StartTime.String(),
```
[the remark this answers is not in the document]
Comment (ChrisSchinnerl) [reply api/prometheus.go:654]: Probably a good idea to double check what the prometheus convention is and then pick that

File: api/utils.go (hunk 2 above)
Comment (ChrisSchinnerl) [reply api/utils.go:11]: The checker can't work with that right now so I'd keep it here for now.

File: alerts/prometheus.go
[the remark this answers is not in the document]
Comment (chris124567) [reply alerts/prometheus.go:1]: How do you want to handle client end of these requests for the test?  Should I just do it in ad hoc way and make an HTTP request helper or should there be a Jape client for the Prometheus endpoints?

File: api/prometheus.go [hunk 5]
```diff
@@ -0,0 +1,869 @@
...
+		"siamux_port": host.Settings.SiaMuxPort,
+	}
+	netAddressLabel := map[string]any{
+		"net_address": host.NetAddress,
+	}
+
+	return []prometheus.Metric{
+		{
+			Name:   "renterd_host_scanned",
+			Labels: netAddressLabel,
```
Comment (peterjan) [inline api/prometheus.go:121]: move down to `interactions` section?

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -25,6 +25,8 @@ jobs:
             autopilot
             bus bus/client
             worker worker/client
+          flags: |
+            -japecheck.types=false
```
Comment (peterjan) [inline .github/workflows/test.yml:29]: This is kind of unfortunate. Do we have to do this or could this be a debug leftover? If we have to do it, can we get around doing it by using `jc.Custom` in the required places.

**Review by peterjan — APPROVED** [review]
looks good to me 👍

File: .github/workflows/test.yml (hunk 6 above)
Comment (n8maninger) [reply .github/workflows/test.yml:29]: `jc.Custom` is pretty annoying and nasty. Better to not use it.

Comment (chris124567) [reply .github/workflows/test.yml]: Oops. Missed this.  I didn't think it was too bad but I'll undo it

Comment (peterjan) [reply .github/workflows/test.yml:29]: Hm ok. I'm a bit surprised we're giving up on type checking only for the sake of same `jc.Custom`'s. I agree they are annoying but at the same time they're not really in the way of anything. Thanks for the link on the discussion, I missed that.

**Review by peterjan — APPROVED** [review]

**Comment by peterjan** [comment]
> Should be OK to merge?

Yes I re-approved. Chris does the merges as he is the merge master. I personally like having one person doing all of the merging to ensure we're not getting in each other's way too much. Now that we're 3 devs contributing though that might become increasingly annoying.

File: .github/workflows/test.yml (hunk 6 above)
Comment (n8maninger) [reply .github/workflows/test.yml:29]: I agree it's a bit sad, but we spend a lot of time and effort fighting jape to do normal API things. Adding no-op funcs to satisfy an external analyzer isn't ideal. You can pretty much get the same results with a test, imo.

**Pull request merged**